File: backend/app/equipment_profiles.py
from __future__ import annotations

from dataclasses import dataclass


GENERIC_IMAGE_PROFILE_KEY = "generic_image"
SEBIA_AGAROSE_IMAGE_PROFILE_KEY = "sebia_agarose_image"
SEBIA_CAPILLARY_IMAGE_PROFILE_KEY = "sebia_capillary_image"

SEBIA_BRAND_TOKEN = "SEBIA"
SEBIA_AGAROSE_MODEL_TOKENS = ("HYDRASYS", "HYDRAGEL")
SEBIA_CAPILLARY_MODEL_TOKENS = ("CAPILLARYS", "MINICAP")


@dataclass(frozen=True)
class EquipmentProfileResolution:
    key: str
    label: str
    origin: str | None
    model: str | None
    uses_sebia_agarose_guardrails: bool = False
    prefers_curve_input: bool = False


def normalize_equipment_text(value: str | None) -> str:
    if not value:
        return ""
    return " ".join(value.strip().upper().split())


def resolve_equipment_profile(origin: str | None, model: str | None) -> EquipmentProfileResolution:
    normalized_origin = normalize_equipment_text(origin)
    normalized_model = normalize_equipment_text(model)
    normalized_combined = " ".join(part for part in (normalized_origin, normalized_model) if part)
    is_sebia = SEBIA_BRAND_TOKEN in normalized_combined

    if any(token in normalized_combined for token in SEBIA_CAPILLARY_MODEL_TOKENS):
        return EquipmentProfileResolution(
            key=SEBIA_CAPILLARY_IMAGE_PROFILE_KEY,
            label="SEBIA capilar (imagen preliminar)",
            origin=origin,
            model=model,
            prefers_curve_input=True,
        )

    if any(token in normalized_combined for token in SEBIA_AGAROSE_MODEL_TOKENS):
        return EquipmentProfileResolution(
            key=SEBIA_AGAROSE_IMAGE_PROFILE_KEY,
            label="SEBIA gel / agarosa",
            origin=origin,
            model=model,
            uses_sebia_agarose_guardrails=True,
        )

    if is_sebia:
        return EquipmentProfileResolution(
            key=SEBIA_AGAROSE_IMAGE_PROFILE_KEY,
            label="SEBIA no especificado (fallback gel/agarosa)",
            origin=origin,
            model=model,
            uses_sebia_agarose_guardrails=True,
        )

    return EquipmentProfileResolution(
        key=GENERIC_IMAGE_PROFILE_KEY,
        label="Perfil generico por imagen",
        origin=origin,
        model=model,
    )

File: src/lib/equipmentProfiles.ts
export type ProcessorEquipmentProfileKey =
  | 'generic_image'
  | 'sebia_agarose_image'
  | 'sebia_capillary_image'

export type ProcessorEquipmentProfile = {
  key: ProcessorEquipmentProfileKey
  label: string
  origin: string | null
  model: string | null
  usesSebiaAgaroseGuardrails: boolean
  prefersCurveInput: boolean
}

const SEBIA_BRAND_TOKEN = 'SEBIA'
const SEBIA_AGAROSE_MODEL_TOKENS = ['HYDRASYS', 'HYDRAGEL']
const SEBIA_CAPILLARY_MODEL_TOKENS = ['CAPILLARYS', 'MINICAP']

export function normalizeEquipmentText(value: string | null | undefined) {
  if (!value) return ''
  return value.trim().toUpperCase().replace(/\s+/g, ' ')
}

export function resolveProcessingEquipmentProfile(
  origin: string | null | undefined,
  model: string | null | undefined,
): ProcessorEquipmentProfile {
  const normalizedOrigin = normalizeEquipmentText(origin)
  const normalizedModel = normalizeEquipmentText(model)
  const combined = [normalizedOrigin, normalizedModel].filter(Boolean).join(' ')
  const isSebia = combined.includes(SEBIA_BRAND_TOKEN)

  if (SEBIA_CAPILLARY_MODEL_TOKENS.some(token => combined.includes(token))) {
    return {
      key: 'sebia_capillary_image',
      label: 'SEBIA capilar (imagen preliminar)',
      origin: origin?.trim() || null,
      model: model?.trim() || null,
      usesSebiaAgaroseGuardrails: false,
      prefersCurveInput: true,
    }
  }

  if (SEBIA_AGAROSE_MODEL_TOKENS.some(token => combined.includes(token))) {
    return {
      key: 'sebia_agarose_image',
      label: 'SEBIA gel / agarosa',
      origin: origin?.trim() || null,
      model: model?.trim() || null,
      usesSebiaAgaroseGuardrails: true,
      prefersCurveInput: false,
    }
  }

  if (isSebia) {
    return {
      key: 'generic_image',
      label: 'SEBIA no especificado (sin guardrails de metodo)',
      origin: origin?.trim() || null,
      model: model?.trim() || null,
      usesSebiaAgaroseGuardrails: false,
      prefersCurveInput: true,
    }
  }

  return {
    key: 'generic_image',
    label: 'Perfil generico por imagen',
    origin: origin?.trim() || null,
    model: model?.trim() || null,
    usesSebiaAgaroseGuardrails: false,
    prefersCurveInput: false,
  }
}
